### PR TITLE
ValidatorSet change delayed by 1 block, and lite refactor

### PR DESCRIPTION
**Pull request by jaekwon**
The transactions in the LastBlock are sufficient for validators to know what the next validator should be, because it can update the validator set based on the validator updates returned by EndBlock.  But light clients don't know that, so if the application wants to change the validator set completely 100%, a general purpose tendermint light client can't validate it because it doesn't know how to interpret the txs which caused the validator changes.

Delaying validator set updates by 1 block allows the next pending validator set to be included in the block header as "NextValidators", so light clients can know how the validator set changed for a blockchain, even with arbitrary validator set changes.

This PR includes updates for package "lite" as well.

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -1525,7 +1524,9 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 		blockID, ok := precommits.TwoThirdsMajority()
 		if ok {
 			if len(blockID.Hash) == 0 {
-				cs.enterNewRound(height, vote.Round+1)
+				cs.enterNewRound(height, vote.Round)
+				cs.enterPrecommit(height, vote.Round)
+				cs.enterPrecommitWait(height, vote.Round)
```
Comment (jaekwon) [inline consensus/state.go]: This could be a separate PR but I made the change here so... LMK if you want me to fork it into a separate PR.

File: consensus/state_test.go [hunk 2]
```diff
@@ -734,7 +736,7 @@ func TestStateLockPOLSafety1(t *testing.T) {
 	rs := re.(types.EventDataRoundState).RoundState.(*cstypes.RoundState)
 	propBlock := rs.ProposalBlock
 
-	<-voteCh // prevote
+	ensureVote(voteCh, h, r, types.VoteTypePrevote)
```
Comment (jaekwon) [inline consensus/state_test.go]: I only did this for this test case because it's the one that broke (awesome btw, tests are working as expected... pubsub or eventbus needs better debug tooling).

ensureVote etc instead of blindly receiving on the channel will help.  For example, it would have helped me tremendously to know that a prevote was expected but a precommit came through.

File: Makefile [hunk 3]
```diff
@@ -135,11 +135,11 @@ vagrant_test:
 ### go tests
 test:
 	@echo "--> Running go test"
-	@go test $(PACKAGES)
+	@GOCACHE=off go test -p 1 $(PACKAGES)
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply Makefile]: It doesn't restrict to 1 CPU afaik.
AFAIK it runs one test at a time, which mitigates issues around timeouts due to tests running too slow due to too many tests running simultaneously.

e.g. test make100 will succeed (except 1 minor issue) on my mac with this, whereas without, I see all kinds of nondeterministic test failures, which I'm sure we're seeing on Circle as well.

[the remark this answers is not in the document]
Comment (melekes) [reply Makefile]: > too slow due to too many tests running simultaneously.

I thought by default it's number of cores available https://golang.org/pkg/runtime/#GOMAXPROCS. So if you have 2 cores, it should be 2 tests running simultaneously.

[the remark this answers is not in the document]
Comment (jaekwon) [reply Makefile]: You should be running with more cores than 1... and some tests test multiple reactors etc and have real-time timeouts (for lack of virtual deterministic time) that would be reasonable in a -p 1 scenario (utilizing potentially many cores if needed by the test), but not reasonable on Circle or my mac.

The target for Tendermint is for multicore machines.  I see no practical problem with setting -p 1... the tests pass fairly quickly anyways.  The deterministic success makes us waste less time trying to figure out why something broke.

[the remark this answers is not in the document]
Comment (melekes) [reply Makefile]: > The deterministic success makes us waste less time trying to figure out why something broke.

fair argument. thank you

**Comment by ebuchman** [comment]
Oh we should also make the linter happy. 

Should we retarget this PR against develop ?

File: lite/client/provider.go [hunk 4]
```diff
@@ -23,119 +23,111 @@ type SignStatusClient interface {
...
+		// TODO pass through other types of errors.
+		return nil, lerr.ErrMissingValidators(chainID, height)
 	}
-	fc := lite.NewFullCommit(
-		CommitFromResult(commit),
-		types.NewValidatorSet(vals.Validators),
-	)
-	return fc, nil
+	valset = types.NewValidatorSet(res.Validators)
+	valset.TotalVotingPower() // to test deep equality.
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply lite/client/provider.go]: not used, removing.

File: lite/dbprovider.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+//----------------------------------------
+
+func signedHeaderKey(chainID string, height int64) []byte {
+	return []byte(fmt.Sprintf("%s/%010d/sh", chainID, height))
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply lite/dbprovider.go]: no, I think it's clear enough like this.  that key is needed for traversing for some functions.

**Comment by cwgoes** [comment]
> Should we retarget this PR against develop?

Yes; my PR was dropped in favor of https://github.com/tendermint/tendermint/pull/1782 (which has most of the same changes).

File: lite/doc.go [hunk 6]
```diff
@@ -1,133 +1,139 @@
...
-unbonding period to avoid the "nothing at stake" problem.
-If you haven't seen the state in a few months, you will need
-to manually verify the new validator set hash using off-chain
-means (the same as getting the initial hash).
+If the validator set for the blockchain is dynamic, verifying block commits is
+a bit more involved -- if there is a block at height H with a known (trusted)
+validator set V, and another block at height H' (H' > H) with validator set V'
+!= V, then we want a way to safely update it.
+
+First, we get the new (unconfirmed) validator set V' and verify that H' is
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply lite/doc.go]: yeah, that's what BaseCertifier does, which InquiringCertifier uses.  The latter constructs the former with a trusted-fc.NextValidator.

File: lite/inquiring_certifier.go [hunk 7]
```diff
@@ -1,163 +1,209 @@
...
+	} else {
+		// Create filled FullCommit.
+		nfc := FullCommit{
+			SignedHeader:   shdr,
+			Validators:     tfc.NextValidators,
+			NextValidators: nextValset,
+		}
+		// Validate the full commit.  This checks the cryptographic
+		// signatures of Commit against Validators.
+		if err := nfc.ValidateBasic(ic.chainID); err != nil {
```
[the remark this answers is not in the document]
Comment (jaekwon) [reply lite/inquiring_certifier.go]: Yeah I wondered too.  I'll call it ValidateFull then.  Or just Validate.  PR will say ValidateFull.

**Comment by melekes** [comment]
Superseded by #1815
